clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -1291,15 +1336,61 @@ raw2trace_t::write(void *tls, const trace_entry_t *start, const trace_entry_t *e
                     return error;
                 start = it;
                 DEBUG_ASSERT(tdata->cur_chunk_instr_count == 0);
-                // TODO i#5724: We need to re-emit encodings for "it" and any further
-                // instrs in this buffer: have a callback passed in which constructs
-                // an encoding from an instr record?
             }
-            if (type_is_instr(static_cast<trace_type_t>(it->type)))
+            if (type_is_instr(static_cast<trace_type_t>(it->type))) {
                 ++tdata->cur_chunk_instr_count;
+                ++instr_ordinal;
+                if (instr_ordinal >= static_cast<int>(decode_pcs.size()))
+                    return "decode_pcs is missing entries for written instructions";
+            }
+            // Check for missing encodings after possibly opening a new chunk.
+            //
+            // XXX i#5724: Could we add a trace_entry_t-level invariant checker to
+            // identify missing post-chunk encodings?  Or should we have the reader
+            // deliberately clear its encoding history on a chunk boundary, raising
+            // a fatal error on a missing encoding?  For now the only complex case
+            // is these already-generated records which we handle here and have a
+            // unit test covering so those further checks are lower priority.
+            if (TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, tdata->file_type) &&
+                type_is_instr(static_cast<trace_type_t>(it->type)) &&
+                // We don't want encodings for the PC-only i-filtered entries.
+                it->size > 0 && !prev_was_encoding &&
+                record_encoding_emitted(tls, decode_pcs[instr_ordinal])) {
```

Comment:
So this logic seems to properly handle branches that were delayed multiple times. Maybe comment about that?

Reply:
Done.